pstoreds/addr_book.go

```diff
@@ -79,141 +98,206 @@ func (mgr *dsAddrBook) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Durati
 	if ttl <= 0 {
 		return
 	}
-
-	mgr.setAddrs(p, addrs, ttl, true)
+	mgr.setAddrs(p, addrs, ttl, false)
 }
 
 // SetAddr will add or update the TTL of an address in the AddrBook.
 func (mgr *dsAddrBook) SetAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
-	mgr.SetAddrs(p, []ma.Multiaddr{addr}, ttl)
+	addrs := []ma.Multiaddr{addr}
+	mgr.SetAddrs(p, addrs, ttl)
 }
 
 // SetAddrs will add or update the TTLs of addresses in the AddrBook.
 func (mgr *dsAddrBook) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
-	mgr.setAddrs(p, addrs, ttl, false)
+	if ttl <= 0 {
```

Comment:
Before, the ttl was checked on every iteration over addrs to decide what to do. Now we only check it once and dispatch to the right behaviour at the top level.